whylog/teacher/rule_validation_problems.py

```diff
@@ -1,110 +1,89 @@
 
 import six
 
-ValidationResult = namedtuple('ValidationResult', ['errors', 'warnings'])
-
 
 class ValidationResult(object):
-    def __init__(self, errors, warnings):
+    def __init__(self, rule_problems, parser_problems, constraint_problems):
         """
-        :param errors: problems preventing form rule saving
-        :param warnings: problems accepted while rule saving
-        :type errors: [RuleValidationProblem]
-        :type warnings: [RuleValidationProblem]
+        :param rule_problems: Problems related to whole rule
+        :param parser_problems: Problems related to parsers, dict[parser id, problem]
+        :param constraint_problems: Problems related to constraints, dict[constraint id, problem]
+        :type rule_problems: list[RuleValidationProblem]
+        :type parser_problems: dict[int, ParserValidationProblem]
+        :type constraint_problems: dict[int, ConstraintValidationProblem]
         """
-        self.errors = errors
-        self.warnings = warnings
+        self.rule_problems = rule_problems
+        self.parser_problems = parser_problems
+        self.constraint_problems = constraint_problems
 
-    @classmethod
-    def result_from_results(cls, results):
-        errors = sum([result.errors for result in results], [])
-        warnings = sum([result.warnings for result in results], [])
-        return ValidationResult(errors, warnings)
+    def is_acceptable(self):
+        parser_problems_list = itertools.chain.from_iterable(six.itervalues(self.parser_problems))
+        constraint_problems_list = itertools.chain.from_iterable(
+            six.itervalues(self.constraint_problems)
+        )  # yapf: disable
 
-    def acceptable(self):
-        return not self.errors
+        all_problems = itertools.chain.from_iterable(
+            [self.rule_problems, parser_problems_list, constraint_problems_list]
+        )
+        return all([not problem.IS_FATAL for problem in all_problems])
 
-    def select_parser_problems(self, line_id):
-        return [
-            problem for problem in itertools.chain(self.warnings, self.errors)
-            if problem.concerns_parser(line_id)
-        ]  # yapf: disable
+    def in_parser_problems(self, parser_id, problem):
+        problems = self.parser_problems.get(parser_id)
+        if problems is None:
+            return False
+        return problem in problems
 
-    def select_constraint_problems(self, constraint_id):
-        return [
-            problem for problem in itertools.chain(self.warnings, self.errors)
-            if problem.concerns_constraint(constraint_id)
-        ]  # yapf: disable
+    def in_rule_problems(self, problem):
+        return problem in self.rule_problems
 
 
-@six.add_metaclass(ABCMeta)
 class RuleValidationProblem(object):
-    def __eq__(self, other):
-        return self.__dict__ == other.__dict__
-
-    @abstractmethod
-    def concerns_parser(self, parser_id):
-        pass
+    MESSAGE_TEMPLATE = 'Rule problem: %s'
```

Comment:
I wanted Problems to have shared header like `Parser problem: MESSAGE`, `Constraint problem: MESSAGE`. But ok, if it is not needed I will remove it as you suggest.